// bitgo/dist/src/v2/coins/ofcToken.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var ofc_1 = require("./ofc");
var lodash_1 = require("lodash");
var Bluebird = require("bluebird");
var co = Bluebird.coroutine;
var publicIdRegex = /^[a-f\d]{32}$/i;
var OfcToken = /** @class */ (function (_super) {
    __extends(OfcToken, _super);
    function OfcToken(bitgo, tokenConfig) {
        var _this = _super.call(this, bitgo) || this;
        _this.tokenConfig = tokenConfig;
        return _this;
    }
    Object.defineProperty(OfcToken.prototype, "coin", {
        get: function () {
            return this.tokenConfig.coin;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OfcToken.prototype, "decimalPlaces", {
        get: function () {
            return this.tokenConfig.decimalPlaces;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OfcToken.prototype, "name", {
        get: function () {
            return this.tokenConfig.name;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OfcToken.prototype, "backingCoin", {
        get: function () {
            return this.tokenConfig.backingCoin;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OfcToken.prototype, "isFiat", {
        get: function () {
            return this.tokenConfig.isFiat;
        },
        enumerable: true,
        configurable: true
    });
    OfcToken.prototype.getChain = function () {
        return this.type;
    };
    OfcToken.prototype.getFullName = function () {
        return this.name;
    };
    OfcToken.prototype.getBaseFactor = function () {
        return String(Math.pow(10, this.decimalPlaces));
    };
    Object.defineProperty(OfcToken.prototype, "type", {
        get: function () {
            return this.tokenConfig.type;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Flag for sending value of 0
     * @returns {boolean} True if okay to send 0 value, false otherwise
     */
    OfcToken.prototype.valuelessTransferAllowed = function () {
        return false;
    };
    OfcToken.createTokenConstructor = function (config) {
        return function (bitgo) { return new OfcToken(bitgo, config); };
    };
    /**
     * Assemble keychain and half-sign prebuilt transaction
     * @param params
     * @param callback
     * @returns {Bluebird<SignedTransaction>}
     */
    OfcToken.prototype.signTransaction = function (params, callback) {
        return co(function () {
            var txPrebuild, payload, signatureBuffer, signature;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        txPrebuild = params.txPrebuild;
                        payload = txPrebuild.payload;
                        return [4 /*yield*/, this.signMessage(params, payload)];
                    case 1:
                        signatureBuffer = _a.sent();
                        signature = signatureBuffer.toString('hex');
                        return [2 /*return*/, { halfSigned: { payload: payload, signature: signature } }];
                }
            });
        })
            .call(this)
            .asCallback(callback);
    };
    /**
     * Check if an address is valid for this ofc token.
     *
     * These addresses are either bg-<publicid>, where public id is the internal address to send to,
     * or are an address which is valid on the backing coin of this ofc token.
     * @param address address to check for validity
     */
    OfcToken.prototype.isValidAddress = function (address) {
        if (!lodash_1.isString(address)) {
            return false;
        }
        if (address.startsWith('bg-')) {
            var parts = address.split('-');
            var accountId = parts[1];
            return parts.length === 2 && publicIdRegex.test(accountId);
        }
        else {
            var backingCoin = this.bitgo.coin(this.backingCoin);
            return backingCoin.isValidAddress(address);
        }
    };
    return OfcToken;
}(ofc_1.Ofc));
exports.OfcToken = OfcToken;
