// bitgo/dist/src/v2/trading/settlement.js
"use strict";
/**
 * @prettier
 */
Object.defineProperty(exports, "__esModule", { value: true });
var affirmation_1 = require("./affirmation");
var SettlementStatus;
(function (SettlementStatus) {
    SettlementStatus["CANCELED"] = "canceled";
    SettlementStatus["PENDING"] = "pending";
    SettlementStatus["REJECTED"] = "rejected";
    SettlementStatus["SETTLED"] = "settled";
    SettlementStatus["FAILED"] = "failed";
})(SettlementStatus = exports.SettlementStatus || (exports.SettlementStatus = {}));
var SettlementType;
(function (SettlementType) {
    SettlementType["DIRECT"] = "direct";
    SettlementType["AGENCY"] = "agency";
})(SettlementType = exports.SettlementType || (exports.SettlementType = {}));
var Settlement = /** @class */ (function () {
    function Settlement(settlementData, bitgo, enterpriseId) {
        var _this = this;
        this.bitgo = bitgo;
        this.enterpriseId = enterpriseId;
        this.id = settlementData.id;
        this.requesterAccountId = settlementData.requesterAccountId;
        this.status = settlementData.status;
        this.type = settlementData.type;
        this.affirmations = settlementData.affirmations.map(function (affirmation) { return new affirmation_1.Affirmation(affirmation, _this.bitgo, _this.enterpriseId); });
        this.createdAt = new Date(settlementData.createdAt);
        this.expireAt = new Date(settlementData.expireAt);
        this.settledAt = new Date(settlementData.settledAt);
        this.trades = settlementData.trades;
    }
    return Settlement;
}());
exports.Settlement = Settlement;
