// bitgo/dist/src/v2/trading/tradingPartners.js
"use strict";
/**
 * @prettier
 */
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 */
var Bluebird = require("bluebird");
var tradingPartner_1 = require("./tradingPartner");
var co = Bluebird.coroutine;
// Side of the requester (if they should be considered the primary or the secondary)
// Only important for agency partnerships
// the primaryAccount is the agent, settling trades for the secondary account id
var TradingReferralRequesterSide;
(function (TradingReferralRequesterSide) {
    TradingReferralRequesterSide["PRIMARY"] = "primary";
    TradingReferralRequesterSide["SECONDARY"] = "secondary";
})(TradingReferralRequesterSide = exports.TradingReferralRequesterSide || (exports.TradingReferralRequesterSide = {}));
var TradingPartners = /** @class */ (function () {
    function TradingPartners(bitgo, enterpriseId, account) {
        this.bitgo = bitgo;
        this.enterpriseId = enterpriseId;
        this.account = account;
    }
    /**
     * List all trading partners of your trading account. Your trading partners are the accounts you are allowed to settle with.
     * @param callback
     */
    TradingPartners.prototype.list = function (callback) {
        var self = this;
        return co(function list() {
            var url, response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        url = self.bitgo.microservicesUrl("/api/trade/v1/enterprise/" + self.enterpriseId + "/account/" + self.account.id + "/tradingpartners");
                        return [4 /*yield*/, self.bitgo.get(url).result()];
                    case 1:
                        response = _a.sent();
                        return [2 /*return*/, response.tradingPartners.map(function (partner) { return new tradingPartner_1.TradingPartner(partner, self.bitgo, self.enterpriseId, self.account); })];
                }
            });
        })
            .call(this)
            .asCallback(callback);
    };
    /**
     * Add trading partner given the unique referralCode provided by trading partner.
     * @param params
     * @param params.referralCode unique referral code provided by counterparty
     * @param params.type type of trading partnership
     * @param params.requesterSide side of the requester (primary or secondary) important for agency relationships
     * @param callback
     */
    TradingPartners.prototype.addByCode = function (params, callback) {
        var self = this;
        return co(function refer() {
            var url, response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        url = self.bitgo.microservicesUrl("/api/trade/v1/enterprise/" + self.enterpriseId + "/account/" + self.account.id + "/tradingpartners");
                        return [4 /*yield*/, self.bitgo
                                .post(url)
                                .send(params)
                                .result()];
                    case 1:
                        response = _a.sent();
                        return [2 /*return*/, response];
                }
            });
        })
            .call(this)
            .asCallback(callback);
    };
    return TradingPartners;
}());
exports.TradingPartners = TradingPartners;
