// bitgo/dist/src/v2/trading/tradingPartner.js
"use strict";
/**
 * @prettier
 */
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 */
var Bluebird = require("bluebird");
var co = Bluebird.coroutine;
var TradingPartnerStatus;
(function (TradingPartnerStatus) {
    TradingPartnerStatus["ACCEPTED"] = "accepted";
    TradingPartnerStatus["REJECTED"] = "rejected";
    TradingPartnerStatus["CANCELED"] = "canceled";
    TradingPartnerStatus["PENDING"] = "pending";
})(TradingPartnerStatus = exports.TradingPartnerStatus || (exports.TradingPartnerStatus = {}));
// Type of partnership used for settlements
var TradingPartnerType;
(function (TradingPartnerType) {
    TradingPartnerType["DIRECT"] = "direct";
    TradingPartnerType["AGENCY"] = "agency";
})(TradingPartnerType = exports.TradingPartnerType || (exports.TradingPartnerType = {}));
var TradingPartner = /** @class */ (function () {
    function TradingPartner(tradingPartnerData, bitgo, enterpriseId, currentAccount) {
        this.id = tradingPartnerData.id;
        this.primaryEnterpriseName = tradingPartnerData.primaryEnterpriseName;
        this.primaryAccountId = tradingPartnerData.primaryAccountId;
        this.secondaryEnterpriseName = tradingPartnerData.secondaryEnterpriseName;
        this.secondaryAccountId = tradingPartnerData.secondaryAccountId;
        this.requesterAccountId = tradingPartnerData.requesterAccountId;
        this.status = tradingPartnerData.status;
        this.type = tradingPartnerData.type;
        this.bitgo = bitgo;
        this.enterpriseId = enterpriseId;
        this.currentAccount = currentAccount;
    }
    /**
     * Check if a trading partner has enough funds to cover the cost of a trade.
     * @param currency the currency to check
     * @param amount the amount of currency to check, represented in base units (such as cents, satoshi, or wei)
     * @param callback
     */
    TradingPartner.prototype.checkBalance = function (currency, amount, callback) {
        var self = this;
        var partnerAccountId = self.primaryAccountId === self.currentAccount.id ? self.secondaryAccountId : self.primaryAccountId;
        return co(function checkBalance() {
            var url, response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        url = self.bitgo.microservicesUrl("/api/trade/v1/enterprise/" + self.enterpriseId + "/account/" + self.currentAccount.id + "/tradingpartners/" + partnerAccountId + "/balance");
                        return [4 /*yield*/, self.bitgo
                                .get(url)
                                .query({ currency: currency, amount: amount })
                                .result()];
                    case 1:
                        response = _a.sent();
                        return [2 /*return*/, response.check];
                }
            });
        })
            .call(this)
            .asCallback(callback);
    };
    return TradingPartner;
}());
exports.TradingPartner = TradingPartner;
